<h1>Содержание проекта<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подключение-к-БД-и-вывод-таблиц" data-toc-modified-id="Подключение-к-БД-и-вывод-таблиц-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подключение к БД и вывод таблиц</a></span></li><li><span><a href="#Вывод-таблиц" data-toc-modified-id="Вывод-таблиц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Вывод таблиц</a></span></li><li><span><a href="#Исследование-базы-данных" data-toc-modified-id="Исследование-базы-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследование базы данных</a></span><ul class="toc-item"><li><span><a href="#Задача-1" data-toc-modified-id="Задача-1-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Задача 1</a></span></li><li><span><a href="#Задача-2" data-toc-modified-id="Задача-2-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Задача 2</a></span></li><li><span><a href="#Задача-3" data-toc-modified-id="Задача-3-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Задача 3</a></span></li><li><span><a href="#Задача-4" data-toc-modified-id="Задача-4-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Задача 4</a></span></li><li><span><a href="#Задача-5" data-toc-modified-id="Задача-5-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Задача 5</a></span></li></ul></li><li><span><a href="#Общий-вывод-по-итогам-исследования" data-toc-modified-id="Общий-вывод-по-итогам-исследования-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод по итогам исследования</a></span></li></ul></div>

# Выпускной проект. SQL

Пандемия коронавируса изменила привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг, и это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Компания решила купить крупный сервис для чтения книг по подписке. Задача проекта — проанализировать базу данных и сформулировать ценностное предложение для нового продукта.

В базе данных компании содержится информация о книгах, издательствах, авторах, а также пользовательские обзоры книг.

In [1]:
# импорт библиотек

# для хранения результатов
import pandas as pd

# для коннекции к БД и выводу таблиц
from sqlalchemy import create_engine

## Подключение к БД и вывод таблиц

In [2]:
# установка параметров и подключение к БД
db_config = {'user': 'praktikum_student',                           # имя пользователя
             'pwd': 'Sfd4$2;e-d30tp',                               # пароль
             'host': 'rt1b-wcoijdj3yxfaf3qs.mdb.yandexcloud.net',   # хост
             'port': 6431,                                          # порт подключения
             'db': 'data-analyst-final-project-db'}                 # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])

# сохранение коннектора
engine = create_engine(connection_string, connect_args = {'sslmode':'require'})

Для будущих выводов результатов SQL-запросов подготовим функцию `sql_request_result` - на вход получает переменную query, содержащую SQL-запрос, затем возвращает результат в pandas

In [3]:
# подготовка функции для вывода результатов запросов
def sql_request_result(query):
    return pd.io.sql.read_sql(query, con = engine)

## Вывод таблиц

Выведем все таблицы, проверим, что все работает корректно. Чтобы не сильно нагружать БД, ограничимся 5 запросами

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

In [4]:
# запрос к таблице books
query = '''

        SELECT *
        FROM books
        LIMIT 5;

        '''

# вывод таблицы books
sql_request_result(query)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

In [5]:
# запрос к таблице authors
query = '''

        SELECT *
        FROM authors
        LIMIT 5;

        '''

# вывод таблицы authors
sql_request_result(query)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства.

In [6]:
# запрос к таблице publishers
query = '''

        SELECT *
        FROM publishers
        LIMIT 5;

        '''

# вывод таблицы publishers
sql_request_result(query)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

In [7]:
# запрос к таблице ratings
query = '''

        SELECT *
        FROM ratings
        LIMIT 5;

        '''

# вывод таблицы ratings
sql_request_result(query)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

In [8]:
# запрос к таблице reviews
query = '''

        SELECT *
        FROM reviews
        LIMIT 5;

        '''

# вывод таблицы reviews
sql_request_result(query)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Запросы отработали корректно, каждая таблица выгружена.

Можно приступать к исследованию.

## Исследование базы данных

### Задача 1

**Сколько книг вышло после 1 января 2000 года?**

Выводом этого запроса предполагается одно поле - количество книг с датой публикации в 2000 году и позже

In [9]:
# запрос к БД
query = '''

        SELECT COUNT(book_id)
        FROM books
        WHERE publication_date > '2000-01-01';

        '''

# вывод запроса
sql_request_result(query)

count
0    819

Всего в библиотеке имеется 819 книг, выпущенных в 2000 году и позже

### Задача 2

**Сколько обзоров проведено на каждую книгу и какая средняя оценка?**

Выводом этого запроса предполагается 3 поля - название книги, количество обзоров по ней и средний рейтинг

In [10]:
# запрос к БД
query = '''

        SELECT b.title,
               COUNT(DISTINCT rv.review_id) AS reviews_count,
               AVG(rt.rating) AS avg_rating
        FROM books AS b
        LEFT OUTER JOIN reviews AS rv ON b.book_id = rv.book_id
        LEFT OUTER JOIN ratings AS rt ON b.book_id = rt.book_id
        GROUP BY b.title
        ORDER BY reviews_count DESC;

        '''

# отсортируем итоговую таблицу по убыванию количества обзоров

# вывод запроса
sql_request_result(query)

title  reviews_count  \
0                                  Memoirs of a Geisha              8   
1                              Twilight (Twilight  #1)              7   
2    Harry Potter and the Chamber of Secrets (Harry...              6   
3                                     The Glass Castle              6   
4                                      Eat  Pray  Love              6   
..                                                 ...            ...   
994    Anne Rice's The Vampire Lestat: A Graphic Novel              0   
995                            The Natural Way to Draw              0   
996   The Cat in the Hat and Other Dr. Seuss Favorites              0   
997                          Essential Tales and Poems              0   
998  Disney's Beauty and the Beast (A Little Golden...              0   

     avg_rating  
0      4.138462  
1      3.662500  
2      4.287500  
3      4.206897  
4      3.395833  
..          ...  
994    3.666667  
995    3.000000  
996    5.000000  
997    4.000000  
998    4.000000  

[999 rows x 3 columns]

Судя по выводу, в базе сервиса книг имеется 999 книг, которым так или иначе выставлена оценка.

Имеем разброс по количеству обзоров от 0 до 8.

### Задача 3

**Какое издательство выпустило наибольшее число книг толще 50 страниц?**

*Так брошюры будут исключены из анализа*

Выводом этого запроса предполагается два поля с названием издательства, которое выпустило наибольшее количество полноценных книг (> 50 страниц) и само количество книг. Во временной таблице получим таблицу с названием издательства и общим количеством выпущенных книг с сортировкой по убыванию, затем выведем только 1 строку в основном запросе (2 вариант - сделать подзапрос полностью во FROM)

In [11]:
# запрос к БД
query = '''

        WITH books_count AS(SELECT pb.publisher AS publisher,
                                   COUNT(DISTINCT b.book_id) AS published_books
                            FROM publishers AS pb
                            LEFT OUTER JOIN books AS b ON pb.publisher_id = b.publisher_id
                            WHERE b.num_pages > 50
                            GROUP BY pb.publisher
                            ORDER BY published_books DESC)
        
        SELECT *
        FROM books_count
        LIMIT 1;

        '''

# вывод запроса
sql_request_result(query)

publisher  published_books
0  Penguin Books               42

Судя по результатам запроса, наибольшее количество полноценных книг выпустило издательство Penguin Books - 42 книги.

### Задача 4

**Какой автор имеет самую высокую среднюю оценку книг?** 

*Учитываются только книги с 50 и более оценками*

Выводом этого запроса предполагается два поля с именем автора, получившим самые высокие оценки своих книг и само значение рейтинга. Сделаем подзапрос во FROM - получим новую таблицу с названием книги, именем ее автора и общей средней оценкой книги с сортировкой по убыванию оценки. Затем из новой таблицы выберем 2 поля с именем автора и общим средним рейтингом по всем книгам.

In [12]:
# запрос к БД
query = '''

        SELECT book_author,
               AVG(authors_ratings.avg_rating) AS avg_author_rating
        FROM (SELECT a.author AS book_author,
                     b.book_id AS book_id,
                     AVG(r.rating) AS avg_rating
              FROM authors AS a
              LEFT OUTER JOIN books AS b ON a.author_id = b.author_id
              LEFT OUTER JOIN ratings AS r ON b.book_id = r.book_id
              GROUP BY a.author, b.book_id
              HAVING COUNT(r.rating) >= 50
              ORDER BY avg_rating DESC) AS authors_ratings
        GROUP BY book_author
        ORDER BY avg_author_rating DESC
              
        LIMIT 1;

        '''

# вывод запроса
sql_request_result(query)

book_author  avg_author_rating
0  J.K. Rowling/Mary GrandPré           4.283844

Судя по результатам запроса, самый высокий средний рейтинг имеет дуэт авторов: J.K. Rowling/Mary GrandPré - он составляет 4,283

### Задача 5

**Какое среднее количество обзоров сделали пользователи, которые поставили больше 50 оценок?**

Результатом этого запроса предполагается одно поле со значением среднего количества обзоров, которые сделали постоянные читатели (больше 50 оценок).

Пойдем последовательно:
- 1) Получим ники пользователей, кто в целом поставил больше 50 оценок
- 2) Получим количество обзоров, которое сделал каждый пользователь, поставивший больше 50 оценок
- 3) Из получившейся временной таблицы посчитаем среднее значение количества обзоров по всем постоянным читателям

In [13]:
# запрос к БД
query = '''
        
        WITH reviews_count AS(SELECT COUNT(review_id) AS users_reviews_count
                              FROM reviews
                              WHERE username IN (SELECT username
                                                 FROM ratings
                                                 GROUP BY username
                                                 HAVING COUNT(rating) > 50)
                              GROUP BY username)
                              
        SELECT AVG(users_reviews_count) AS avg_reviews_count
        FROM reviews_count;

        '''

# вывод запроса
sql_request_result(query)

avg_reviews_count
0          24.333333

Судя по результатам вывода, в среднем постоянные читатели пишут по 24 обзора. 

Это точно давние пользователи сервиса. Допустим, если за год прочитать 24 книги - на прочтение одной книги уйдет по 15 дней, при этом каждая из них может быть как большой, так и объемной (в этой задаче не учитывалось количество страниц). Может быть и такое, что обзоры написаны не на все прочитанные книги - прочитанных может быть гораздо больше.

## Общий вывод по итогам исследования

В рамках проекта произведено исследование имеющейся базы книжного сервиса. Основные наблюдения:
- 1) В базе имеется 819 книг, опубликованных с начала 2000 года
- 2) Не на все книги выпущены обзоры - разброс составляет от 0 до 8 обзоров
- 3) Наибольшее количество полноценных книг (более 50 страниц) имеется от издательства Penguin Books - всего 42 книги в базе
- 4) Самый высокий средний рейтинг от читателей имеет дуэт авторов J.K. Rowling/Mary GrandPré - он составляет 4,28
- 5) В среднем, давние и постоянные читатели пишут 24 обзора на книги